### PR TITLE
#5072 - Onboarding phase 2 slice 1 - Activate blueprints in the sidebar

### LINKED ISSUE #5072: Slice 1a: Activate on top of the Marketplace page (no blueprint options)
Slice 1a of <https://www.notion.so/pixiebrix/Phase-2-Improve-Onboard-via-Marketplace-Blueprint-5229f05895914c578cb07d49c67fc485>

<p></p>

For this slice, we are assuming: <span data-preserve-white-space></span>

-
    - Already registered\/logged in

    - Already have PB Extension installed

    - No blueprint options are required

  <p></p>

- For all other cases, we will continue to use the current path of activating via the  Extension Console\.

- When a user clicks “Activate” in the Marketplace \(either from the blueprint card or on the blueprint page\), we want to retain the context of the Marketplace by not opening the Extension Console or moving away from the Marketplace\.

<p></p>

<p></p>

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Closes #5072 

## Work left to do
- [x] Check if the user is logged in
- [x] Load the recipe and marketplace listing in the sidebar panel component
- [x] Set up the "finish" view with "Well done!" etc. copy and finish button based on whether or not the blueprint has an example URL
- [x] Show activate options view if the blueprint has options/inputs
- [x] Wire up blueprint activation to happen when the finish view is shown, either immediately or after submitting options/inputs
- [x] Implement `getInProgressRecipeActivation()` function in `marketplace.ts` based on what Aleksandr ends up implementing for slice 2 (probably read persisted redux state directly from local storage)
- [x] Remove debug `notify.info()` calls

## Reviewer Tips

- Check out the `marketplace.ts` script for the entry point to the marketplace page modifications
- The main, new sidebar panel content is in `ActivateRecipePanel`
- The rest of the changes are mostly just wiring up new `sidebarController` and messaging framework methods to connect the marketplace script to the PB sidebar

## Demo

Activation for a not logged in user:
https://www.loom.com/share/8558f23c4aa14a6d8fc6b4abdc68f104

Activation without login or options:
https://www.loom.com/share/20e4fbb61c484001b91916492f306f7e

## Checklist

- [x] Add tests
- [ ] Designate a primary reviewer

### CODE REVIEW

File: src/contentScript/marketplace.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    } catch {
+      continue;
+    }
+
+    // Check if recipe is already activated
+    if (installedRecipeIds.includes(recipeId)) {
+      button.innerHTML = '<i class="fas fa-sync-alt"></i> Reactivate';
+    }
+
+    button.addEventListener("click", async (event) => {
```
Comment (twschiller) [inline src/contentScript/marketplace.ts:111]: Since this runs on each focus, I think these addEventListeners could stack up?

Comment (BLoe) [reply src/contentScript/marketplace.ts:113]: This problem should be prevented by the check on the `enhancementsLoaded` flag, which prevents it from running again on a page. I wasn't able to see/produce duplicate handlers in my dev build.

File: src/contentScript/marketplace.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+export async function initMarketplaceEnhancements() {
+  if (
+    !startsWith(window.location.href, "https://www.pixiebrix.com/marketplace")
+  ) {
+    return;
+  }
+
+  window.addEventListener("focus", async () => {
+    notify.info("Marketplace tab focused");
```
Comment (BALEHOK) [inline src/contentScript/marketplace.ts:144]: See Q re `notify` above

File: src/contentScript/marketplace.test.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+      options?: boolean | AddEventListenerOptions
+    ) {
+      // Store listener reference so it can be removed during reset
+      refs.push({ type, listener, options });
+      // Call original window.addEventListener
+      originalAddEventListener(type, listener, options);
+    }
+
+    global.window.addEventListener = addEventListenerSpy;
+  });
```
[the remark this answers is not in the document]
Comment (BLoe) [reply src/contentScript/marketplace.test.ts]: Yes I was just testing things trying to get this to work, I'm removing all of this code in both places

File: src/contentScript/marketplace.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  controller.signal.addEventListener("abort", () => {
+    hideActivateRecipeInSidebar(recipeId);
+  });
+}
+
+let enhancementsLoaded = false;
+
+async function loadPageEnhancements(): Promise<void> {
+  if (enhancementsLoaded) {
+    notify.info("Enhancements already loaded, skipping");
```
[the remark this answers is not in the document]
Comment (BLoe) [reply src/contentScript/marketplace.ts:93]: I have a step in my "work left to do" section to remove this from the code. Keep in mind this is still a draft PR 😅

File: src/options/pages/marketplace/ServicesBody.tsx [hunk 5]
```diff
@@ -28,15 +28,33 @@ import { type ServiceAuthPair } from "@/core";
...
 interface OwnProps {
   blueprint: RecipeDefinition;
 }
 
+const ErrorComponent: React.FC<{
+  error: string | string[] | undefined;
+  fieldIndex: number;
+}> = ({ error, fieldIndex }) => {
+  if (!Array.isArray(error)) {
+    return <></>;
```
Comment (BALEHOK) [inline src/options/pages/marketplace/ServicesBody.tsx:42]: This looks a little bit weird.

```suggestion
    return null;
```

File: src/options/pages/marketplace/ServicesBody.tsx [hunk 6]
```diff
@@ -28,15 +28,33 @@ import { type ServiceAuthPair } from "@/core";
...
+  }
+
+  // eslint-disable-next-line security/detect-object-injection -- index
+  const fieldError = error[fieldIndex];
+  if (typeof fieldError !== "string") {
+    return <></>;
```
Comment (BALEHOK) [inline src/options/pages/marketplace/ServicesBody.tsx:48]:
```suggestion
    return null;
```

**Review by BALEHOK — COMMENTED** [review]
I see we're adding to the content script code that may pull relatively heavy libraries. How concerned are we about the size of the content script? Should we compare the difference?

File: src/contentScript/marketplace.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+  await loadPageEnhancements();
+
+  if (!(await isUserLoggedIn())) {
+    return;
+  }
+
+  const recipeId = await getInProgressRecipeActivation();
+  if (recipeId) {
+    await showSidebarActivationForRecipe(recipeId);
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/contentScript/marketplace.ts:159]: done

File: src/sidebar/activateRecipe/ActivateRecipePanel.tsx [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    recipeError,
+  ]);
+
+  if (isLoadingRecipe || isLoadingListing) {
+    return <Loader />;
+  }
+
+  if (recipe == null) {
+    throw new Error(`Recipe ${recipeId} not found`);
+  }
```
Comment (BALEHOK) [inline src/sidebar/activateRecipe/ActivateRecipePanel.tsx:123]: If recipe hasn't been found, let user know

**Review by twschiller — APPROVED** [review]

**Comment by twschiller** [comment]
Marking as approved so we can cut a release build for manual QA

**Comment by BLoe** [comment]
@BALEHOK I'm merging this now so we can cut a release build, but feel free to look at the newer commits from me and comment here if you have feedback. I can address it in the follow up PR.

**Pull request merged**